### ISSUE TITLE
FAQ: How to track latest nightly builds

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#11068]
### What is a nightly KOReader release?

KOReader development cycle includes two major milestones: **stable** monthly-ish release suitable for an average _user_, and intermediate **nightly** releases which represent cutting edge features intended mostly for _developers_ and _testers_.

1) Stable monthly-ish release is named `latest release` in GitHub. They have the nomenclature of `vYYYY.MM`, where `YYYY` is the year, and `MM` is the month of the release. For example, current one is `v2023.10`.

(Occasionally there are critical bugs found right after the stable release, then the latest release receives an update like `vYYYY.MM.F`, where `F` is the iteration of fixes. E.g., `v2023.06.1`.)

All stable KOReader releases can be found here:
https://github.com/koreader/koreader/releases

2) In the same vein as Rome wasn't built in a day, KOReader stable releases aren't compiled in a single GitHub commit. Throughout the month after the latest release, many devs put their hard work and passion to improve features, fix bugs, optimize memory, and much more - all under careful inspection of the code and user needs.

### Nightly build compilation mechanism
If there is a newly merged commit, or a Pull Request which contains multiple commits, KOReader development's backend server compiles the so-called `nightly release` (or just `nightly`), around 6-7 AM CEST the closest morning after the merge. They have the nomenclature of `vYYYY.MM-ii-buildnumber_date`, where `ii` is the number of merged commits. After each stable release, the `ii` counter resets.

### Fantastic commits and where to find them
The order and contents of the merged commits that made it to nightlies can be found at:
https://github.com/koreader/koreader/compare/v2023.10...master

After each new stable release the link should be changed accordingly if you would like to track the changes in terms of merged commits/nightlies.

Here, for the latest compiled nightly, the index `ii` will be the number of commits on this page, e.g., `v2023.10-2` for two merged commits.

### Where to download nightlies
All those nightlies can be found at:
http://build.koreader.rocks/download/nightly/

This is the FTP-based approach, where each directory represents a nightly release, and the directories contain all the compiled packages for platforms that KOReader supports.

**That means, if you are waiting for a particular fix, you need to wait about a day until the nightly with `ii` ≥ `number of commits you need` is compiled.**

Nightly releases are intended for a narrow group of testers and developers, so there might some nightlies that are less stable than other nightlies.

### Example
GitHub `compare` shows that by the time of this screenshot there have been 7 commits merged with current `master` branch, relative to the `v2023.10` release:
<img src="https://github.com/koreader/koreader/assets/6279855/37d6f012-963e-45f1-b05d-a4aac3d45c77" height="400">

The list of `nightly` builds, however, shows that the last nightly build contains only 5 commits:
<img src="https://github.com/koreader/koreader/assets/6279855/f304f6fc-6191-45a9-b065-908d838d7077" height="300">

That means that the last 2 commits out of these 7 did not make it to a nightly build yet.
Good news is that you will only need to wait a couple of hours until 6-7 AM CET to be able to download the nightly and test these 2 commits.

##### Nightlies History
Please keep in mind that due to the server-side space limitations there are only a certain number of nightlies that can be concurrently stored on the `koreader.rocks` server. If you intend to test come PRs extensively, makes sure to backup necessary nightlies. 

You can always revert to a closest (or any other) stable release, if you need to.

(System-specific downgrade limitations still apply. For example, for downgrading on Android, you need to delete the newer nightly before you can installer an earlier stable/nightly release).

##### Larger archive on GitLab (4 weeks)

There's a larger archive on https://gitlab.com/koreader/nightly-builds/-/pipelines

(Again, some system-specific limitations apply. For example, for the Android builds specifically there's the minor niggle that you'd have to sign them yourself. That means, currently installed KOReader must be deleted for either downgrades or upgrades w.r.t. the Larger archive nightly builds because the apk signature is now different.)

**Comment by mergen3107** [comment]
@NiLuJe @poire-z @Frenzie @pazos @offset-torque (and everyone interested in this kind of documentation)

Could you please review this piece I came up with? Throughout some months I realized that I need to explain such things about nightlies from scratch to people on forums (4PDA, MobileRead, reddit) and here on GitHub. 

So, I propose to edit this documentation piece and pin it to the of the issues page.

Any edits are welcome (including complete shutdown of such an idea if you deem it unnecessary/useless :D )

We might even consider putting this to the User Guide?

This is my small contribution for making KOReader friendlier to people. I just realized that KOReader is not great because it is just a final product, but because it is also a process! Specifically, a learning process, to which people tend to stick like moths at night stick around the burning lights.

So, this is just another way to support this learning experience. 

Thank you!

**Comment by benoit-pierre** [comment]
> If there is a newly merged Pull Request (PR), KOReader development's backend server compiles the so-called `nightly release` (or just `the nightly`), around 6-7 AM CEST the closest morning after the merge. They have the nomenclature of `vYYYY.MM-II`, where `II` is the number of merged PRs.

Number of additional commits since the last release (you got it right in the example).

**Comment by mergen3107** [comment]
Thank you! @benoit-pierre 
I was hesitant that I missed something like a group PR thingy, but wasn't sure how to fix it exactly 😅

**Comment by Frenzie** [comment]
> Please keep in mind that due to the server-side space limitations there are only a certain number of nightlies that can be concurrently stored on the server. If you intend to test come PRs extensively, makes sure to backup necessary nightlies.

While I don't know to what extent we should advertise it (not because it's secret, but just to prevent confusion), there's a much larger archive on https://gitlab.com/koreader/nightly-builds/-/jobs

For the Android builds specifically there's the minor niggle that you'd have to sign them yourself.

**Comment by mergen3107** [comment]
Thanks @Frenzie !

Interesting, I learned something new again!

I was thinking all along, that even though this FAQ piece is advanced by itself, there probably might also be the `Advanced troubleshooting` as well.

I just didn't know where to draw the line, but your suggestion definitely goes to Advanced, I think.

**Comment by mergen3107** [comment]
I also got a question:
If a commit merging _happens_ to occur right during the compilation time around 6-7 AM CEST, does that mean that first part of the packages will rollout without this new commit? Or does the compilation takes only N number of commits to work with until it's done?

**Comment by Frenzie** [comment]
> does that mean that first part of the packages will rollout without this new commit?

That's theoretically possible, but all builds normally start within seconds of each other.

**Comment by mergen3107** [comment]
I separated the OP into a few more sections. Please have a look and let me know if I need to change one _particular_ heading ;)

**Comment by Frenzie** [comment]
Concretely, we have a script running on the koreader.rocks server that triggers a pipeline at 6 in the morning.

https://github.com/koreader/koreader-misc/blob/c82da186ea90fcd8b76512de4ea683c2503eb72d/nightswatcher/nightswatcher.py#L64-L67

During the European day and especially a few hours later when it overlaps with the American day there's a decent chance that individual jobs might have to wait for others to finish first, but at 6 AM all 15 jobs can start all at once.

Basically the scenario you sketch might be doable right this moment since the American working day is still in full swing, but at 6 the window of opportunity is just so tiny.

But even if it had a decent chance of happening I'd probably hardly consider it an issue.

**Comment by Frenzie** [comment]
Incidentally, out of curiosity and to get https://github.com/koreader/koreader/pull/11069 out in the wild sooner I started a manual pipeline 15 minutes ago and that illustrated what it looks like when they don't start all at once.
https://gitlab.com/koreader/nightly-builds/-/pipelines/1059472074
<img src=https://github.com/koreader/koreader/assets/202757/1e0f96ea-c1e9-42b0-bbd7-2e3245041d6a width=35%>

**Comment by mergen3107** [comment]
Thanks! 

I didn't realize all target platform packages start in parallel!

I was under the impression they are built one by one, don't know why.

**Comment by poire-z** [comment]
(Nothing much to add or correct - it's nice that this is written down and available.)

**Comment by mergen3107** [comment]
Added an example section to clearly show how to find commits/nightlies :D

**Comment by offset-torque** [comment]
> We might even consider putting this to the User Guide?

User guide is essentially aimed at teaching average users how to use KOReader's features. Currently our most complicated procedure in the guide is **"downloading ready made user patches to your device"**.

Quoting your description: **"Nightly releases are intended for a narrow group of testers and developers"**.
So I think this information belongs to the **Developer documentation** instead of the user guide. Because once a user downloads a nightly version, there is nothing else in the user guide that would help them with the next steps. They already have to continue from the Developer documentation.

Of course if our devs think this would be useful in the user guide, I can add it.

**Comment by Frenzie** [comment]
No, I agree that's out of scope. It could potentially make sense to mention in the issue template (i.e., maybe it's already been fixed) but that should also refrain from being too overwhelming.

**Comment by Frenzie** [comment]
Well, there goes any release. ;-)

**Comment by mergen3107** [comment]
I just missed the first PR in the nightly that fixes docsettings/hashsettings, lol :D 
I'll try to patch it up in VM.

**Comment by mergen3107** [comment]
> I'll try to patch it up in VM.

So I tried to diff-patch the kindlepw2 stable 2023.10 package and it complained about spec/ folder missing, which is included in the dev source. 

So, I eneded up... compiling a fresh kindlepw2 nightly myself using my existing dev env :D (plus I got to learn what kox and x-tools are ;) I just did `./kodev release kindlepw2`.

However, in the end in spat out an archive with `targz` in the end. I renamed it to `tar.gz` and was able to extract it.
Is it just my Ubuntu acting up, or I just found a bug?

**Comment by Frenzie** [comment]
Neither, it has something to do with Chinese caching proxies but it's
probably outdated.

**Comment by mergen3107** [comment]
@pazos
What’s the reason you unpinned this issue?
Isn’t it helpful?

**Comment by pazos** [comment]
Fat fingers :/

**Comment by mergen3107** [comment]
Thank you :)

**Comment by shuvashish76** [comment]
Hello, I want to try out Nightly releases.
1. Can I install Stable along with Nightly on the same device? (Android, Linux or other platforms)
2. Will it re-write all my configurations stored for Stable version or they're store separately on a new location? For example for android it's `/storage/emulated/0/koreader` and for linux it's `~/.config/koreader`. What would be the location for Nightly?

**Comment by benoit-pierre** [comment]
On Android, yes, but both versions (F-Droid and the official one) store their settings in the same place.

**Comment by benoit-pierre** [comment]
Same with AppImages on Linux.

**Comment by shuvashish76** [comment]
> On Android, yes, but both versions (F-Droid and the official one) store their settings in the same place.

I guess I've to try with work profile feature apps like [Island](https://f-droid.org/packages/com.oasisfeng.island.fdroid/) or [Shelter](https://f-droid.org/packages/net.typeblog.shelter/) not to mess with stable release setting.

**Comment by shuvashish76** [comment]
Yep it works as I intended :) Now I can test Nightly without messing my stable configurations :D

**Comment by benoit-pierre** [comment]
Does that include different storage spaces so both versions don't write to `/sdcard/koreader`?

**Comment by shuvashish76** [comment]
> Does that include different storage spaces so both versions don't write to `/sdcard/koreader`?

Yes, these apps use Work Profile feature of android to provide isolated spaces.

**Comment by shuvashish76** [comment]
If I open/read books from both versions (not at the same time) is that going to affect any unintended behavior for those `lua` files located inside the `sdr` directory for specific books? Or the Nightly changes are always compatible with stable? e.g. Nightly can write color highlights into the `lua` file but that's not supported in stable version.

**Comment by Frenzie** [comment]
Definitely not always.

**Comment by NiLuJe** [comment]
i.e., it depends ;).

We don't actively try to break stuff, but we also absolutely do not guarantee backward compatibility in any way, shape, or form, be that for main settings of docsettings (i.e., sidecars).

Actual breakages are fairly rare, though, and are likely to be mostly silent (i.e., affect behavior in subtle ways, rather than just plain crashing or something).

**Comment by pazos** [comment]
> KOReader development cycle includes two major milestones: stable monthly-ish release suitable for an average user, and intermediate nightly releases which represent cutting edge features intended mostly for developers and testers.

Not sure how to write all this in a simple way that can be understood.

1. KOReader development happens here on gitub.
2. Each time there's a change on the master branch a nightly is released.
3. From time to time @Frenzie tags a release that gets a name and a version
4. We rarely release fixes for a tagged release
5. Big changes usually happen **after** a tagged release.
6. Big changes usually introduce features and maybe bugs
7. We would like to have all our users using the latest nightly (ie: in sync with master branch) but we understand users on older versions (even n years old!)
8. When a release has some time (ie: +1-2 months) usually no bigger changes are still pending.
9. **If** bigger changes are still pending that means there're no new tagged releases on the horizon.
10. If there're no tagged releases in a while we advice users to switch to the nightly channel.
11. If you want to report a bug you should be running the latest nightly.

I think it describes what happens with this repo.

The original message discourages "average users" to stay in sync with the repo. This, at great scale, causes radio silence the weeks before a release and a few bugs reports after it.

With more users on nightlies all these bug reports will happen before a release, and "stable" releases would be more stable (as in no regresions since last tagged)

**Comment by mergen3107** [comment]
> The original message discourages "average users" to stay in sync with the repo. This, at great scale, causes radio silence the weeks before a release and a few bugs reports after it.

Again, those average users won’t even make it to Github. I mean those who don’t use Style Tweaks, multiple dictionaries, various formats and all. They just read 1937462 books a month, each of which are poorly formatted :D and don’t care. 

Radio silence issue might be fixed with something like:
a) Popup notifications about current nightlies/new features/bug fixes targeting users as soon as they connect to internet :D Evil, I know, but in this context will fix the awareness issues. 

b) adding GitHub on the first page of Quick Start Menu and User Guide. Like, “REPORT BUG ON GITHUB: LINK”

**Comment by mergen3107** [comment]
> and "stable" releases would be more stable (as in no regresions since last tagged)

Unfortunately, this is still a dream. 
See https://github.com/koreader/koreader/issues/12339

The problem, or rather context, of KOReader is that there are so many features that it is physically impossible to test them of. Many devices, many newly introduced features that might or might not break other features. A good example is https://github.com/koreader/koreader/issues/13507

So unless you ask a bunch of beta testers who has all sorts of devices and features, it is hard to predict what's going to happen at each nightly.

**Comment by pazos** [comment]
> Again, those average users won’t even make it to Github. I mean those who don’t use Style Tweaks, multiple dictionaries, various formats and all. They just read 1937462 books a month, each of which are poorly formatted :D and don’t care.

Agree, they spam reddit et al asking the same question again and again. Which, in the best case, leds to other users pointing to this ticket where we say: "don't use nightlies unless you know what you're doing"

While we should say: "nightlies are the way to keep (you and us) updated, stables usually contains more bugs. You're free to opt-out from nightlies as long as you don't experience any bug but you're required to update to last nightly before reporting any"

> So unless you ask a bunch of beta testers who has all sorts of devices and features, it is hard to predict what's going to happen at each nightly.

That's exactly what happens now, after a release, because people have this magic feeling that a tagged release is "stable" and a nightly "might have bugs". So they delay upgrading until tagged releases just to figure out something is now broken in a particular platform-feature-device combo than nobody tested.

The only place where it does make some sense to stay out of nightlies is **after a new release, for a few weeks, sometimes**. Also, staying at nightlies doesn't mean update every day or every week or every month or every year.

You can very much stay in nightlies with a 3 year old install as long as it works for you, but when you upgrade you upgrade to the latest version and not to a version that might be several months old.

**Comment by mergen3107** [comment]
> tagged release is "stable" and a nightly "might have bugs". 

How about we rename "stable" to a, I don't know, "periodic release", or a snapshot release, or something. 

I agree, people tend to think about the stable version as bulletproof one :D

**Comment by offset-torque** [comment]
> **pazos** - ...people have this magic feeling that a tagged release is "stable" and a nightly "might have bugs". So they delay upgrading until tagged releases...

I was in this group until moments ago. But apparently KOReader works more like a rolling linux distro rather than a fixed-release one.

---
### What I can do for the guide

**We have a **"Nightly Versions"** heading under the **Troubleshooting** section in the guide.**
1. I can modify this heading by adding the 11 bullet points from **pazos**'s message.
2. And then add the recommendations from the next message like:

> While we should say: "nightlies are the way to keep (you and us) updated, stables usually contains more bugs. You're free to opt-out from nightlies as long as you don't experience any bug but you're required to update to last nightly before reporting any"

> The only place where it does make some sense to stay out of nightlies is after a new release, for a few weeks, sometimes. Also, staying at nightlies doesn't mean update every day or every week or every month or every year.

3. And then rename this section to **"Which version should I use?"** and move it to the top of the guide as the first heading.
4. Also modify the **"reporting problems"** section to include "you're required to update to last nightly before reporting any bugs".

This section can be linked directly from discussions to avoid repetition of the same infos.

---

> **mergen3107** - How about we rename "stable" to a, I don't know, "periodic release", or a snapshot release, or something.

That name looks misleading in this case. Calling it "stable" automatically and unconsciously labels other releases as non-stable. So a more neutral term can prevent the negative stigma of nightlies.

**Comment by shuvashish76** [comment]
> a) Popup notifications about current nightlies/new features/bug fixes targeting users as soon as they connect to internet :D Evil, I know, but in this context will fix the awareness issues.

Checking updates without user knowledge or permission is considered as [Tracking](https://f-droid.org/en/docs/Anti-Features/#Tracking) Anti-Feature. Do remember to disable it for F-Droid if we go by this route.

> While we should say: "nightlies are the way to keep (you and us) updated, stables usually contains more bugs. You're free to opt-out from nightlies as long as you don't experience any bug but you're required to update to last nightly before reporting any"

Agree with @mergen3107 "nightly" just the name is enough for average users not to consider _development_ version, even if it has less bugs than stable release. I don't think we can ever convince average users to use nightly releases.
Do you think it's better to make monthly release model instead? (I guess for the same reason, Calibre has similar model with stable releases on every alternate Friday.)

**Comment by Frenzie** [comment]
> That name looks misleading in this case. Calling it "stable" automatically and unconsciously labels other releases as non-stable.

That's kind of how it is though. Nightlies have changes that have been through less testing. They might not even start properly, typically due to some combination of settings. Unless we're talking about right this very moment, when the nightlies are simply the upcoming stable give or take some final bugs. (Release candidates, but informally.)

----

> Do you think it's better to make monthly release model instead?

We already have a (bi)monthly release model. When it's not there are basically two reasons:

1. It's not ready, like in January/February.
2. I have too many other things going on, like in December/March.

Granted, a lazier release (GH's newish "generate changelog" feature) is an option, though that doesn't get around the testing.

> (I guess for the same reason, Calibre has similar model with stable releases on every alternate Friday.)

Sounds exhausting. ;-)

**Comment by mergen3107** [comment]
Another example of users thinking it is more "stable" than nightly: https://github.com/koreader/koreader/issues/13536#issuecomment-2789626645

**Comment by offset-torque** [comment]
Wait, so if this is true:

> Frenzie - Nightlies have changes that have been through less testing. They might not even start properly, typically due to some combination of settings. **Unless we're talking about right this very moment, when the nightlies are simply the upcoming stable give or take some final bugs.** (Release candidates, but informally.)

And also this is true:

> Pazos - The only place where it does make some sense to stay out of nightlies is **after a new release, for a few weeks**, sometimes.

It means that stability-green (avoidance/danger zone-red) curve is roughly like this (not to scale :). Nightlies closer to release gets better but just after is nope. Which shows that stable release is actually the "most stable" point:

![Image](https://github.com/user-attachments/assets/bd93af18-b9bf-40d1-bf5d-ed215feebdad)

According to this, best strategy for an average user who just want to read their books is updating their KOReader as close to the peak which is at the STABLE release. But also we want them to test the nightlies in the green zone before STABLE release to catch more bugs for a more "stable" stable release. 

Antyhing wrong here?

**Comment by mergen3107** [comment]
> Anything wrong here?

Nothing wrong in the description, I think this is the way it was designed from the beginning. 

However, reality many times is different. There are points releases (yyyy.mm.f, where f is 01 or 1) that fix critical issues with the stable that just released. This also goes to pazos' point that such things aren't revealed until users install the "stable" release that they think is better.

**Comment by Frenzie** [comment]
> Antyhing wrong here? 

That's pretty much how I see it, yes, a bit more sawtoothed in practice. And it's only showing the top 10% or so.

**Comment by JustinMeena** [comment]
issue Aa raha hai

**Comment by sabmes22** [comment]
- Hello 👋👋 Please, developers, update the Android number to match Android 15 and 16![Screenshot_20251115_043856_F-Droid.png](https://github.com/user-attachments/assets/ea4e6338-bc93-4c55-8b7e-880702073893)

**Comment by alva-seal** [comment]
Are the nightlys still updated? 

I just wondered because the change from the 4th of january are not in a nightly yet?

**Comment by Frenzie** [comment]
Looks like the disk space was used up, surprisingly with the same config that's been running for the past year (keeping 20 old nightlies). I've reduced it to 15 for the moment.

Btw, you can always find all the builds at https://gitlab.com/koreader/nightly-builds/-/jobs?kind=BUILD (and Lua changes don't require any build).

**Comment by alva-seal** [comment]
thank you 👍

**Comment by shuvashish76** [comment]
> If there is a newly merged commit, or a Pull Request which contains multiple commits, KOReader development's backend server compiles the so-called nightly release (or just nightly), around **6-7 AM CEST** the closest morning after the merge. 

@Frenzie Commits are getting merged but no nightly updates since v2026.03-61-g240f591d3_2026-04-24.

**Comment by Frenzie** [comment]
Thanks. It seems the system rebooted five days ago and something went wrong as a consequence (logs simply stopped five days ago…) even though the relevant service did start back up and was running. Maybe it's not very error resistant if the network isn't available during startup or something but I'm not really sure what might've happened there. Anyway, it's running properly now.

**Comment by mergen3107** [comment]
Since main server is down, I resorted to Gitlabs to see what currently nightly has been built. 

I noticed that linux armhf build fails every now and then:
![image](https://github.com/user-attachments/assets/5e091a6d-6c29-4b70-a13d-a1b0f36ebcf0)

Is this an issue?

**Comment by mergen3107** [comment]
Can’t copy from the website on mobile, but here is a screenshot of the failure:

<img width="1290" height="811" alt="Image" src="https://github.com/user-attachments/assets/6672737b-fd8f-4171-be9b-f6b5655f87a9" />

**Comment by Frenzie** [comment]
> Is this an issue?

Kind of, I suppose it should be tracked in virdevenv.

**Comment by alva-seal** [comment]
Are the nightlys somehow down? the primary ota server (ota.koreader.rocks) gives a 522 and the url’s of the others i can choose in the koreader menu seem to have the  version v2026.07.2-75 as the newest from the 18th of august.

Maybe im doing something wrong but thats what i can see

**Comment by benoit-pierre** [comment]
@alva-seal: yes, see #15902.

**Comment by mergen3107** [comment]
You can download nightlies in Gitlab link in the OP if not on Android
